napari/layers/_shapes_layer/model.py

```diff
@@ -0,0 +1,1521 @@
+                shape_type = 'path'
+                self.data.shapes[index].shape_type = shape_type
+            elif shape_type == 'rectangle':
+                # Adding vertex to rectangle turns it into a polygon
+                shape_type = 'polygon'
+                self.data.shapes[index].shape_type = shape_type
+            closed = shape_type != 'path'
+            vertices = self.data._vertices[self.data._index == index]
+            if closed is not True:
+                if int(ind) == 1 and loc < 0:
+                    ind = 0
+                elif int(ind) == len(vertices)-1 and loc > 1:
+                    ind = ind + 1
+
+            vertices = np.insert(vertices, ind, [coord], axis=0)
+            with self.freeze_refresh():
+                self.data.edit(index, vertices)
+                self._selected_box = self.bounding_box(self.selected_shapes)
+            shape, vertex = self._shape_at(coord)
+            self._hover_shape = shape
+            self._hover_vertex = vertex
+            self.refresh()
+            self.status = self.get_message(coord, shape, vertex)
+        elif self.mode == 'vertex_remove':
+            shape, vertex = self._shape_at(coord)
+            if vertex is not None:
+                # have clicked on a current vertex so remove
+                index = shape
+                shape_type = self.data.shapes[index].shape_type
+                if shape_type == 'ellipse':
+                    # Removing vertex from ellipse not implemented
+                    return
+                vertices = self.data._vertices[self.data._index == index]
+                if len(vertices) <= 2:
+                    # If only 2 vertices present, remove whole shape
+                    with self.freeze_refresh():
+                        self.data.remove(index)
+                elif shape_type == 'polygon' and len(vertices) == 3:
+                    # If only 3 vertices of a polygon present remove
+                    with self.freeze_refresh():
+                        self.data.remove(index)
+                else:
+                    if shape_type == 'rectangle':
+                        # Deleting vertex from a rectangle creates a polygon
+                        shape_type = 'polygon'
+                        self.data.shapes[index].shape_type = shape_type
+                    # Remove clicked on vertex
+                    vertices = np.delete(vertices, vertex, axis=0)
+                    with self.freeze_refresh():
+                        self.data.edit(index, vertices)
+                        shapes = self.selected_shapes
+                        self._selected_box = self.bounding_box(shapes)
+                shape, vertex = self._shape_at(coord)
+                self._hover_shape = shape
+                self._hover_vertex = vertex
+                self.refresh()
+                self.status = self.get_message(coord, shape, vertex)
+        else:
+            raise ValueError("Mode not recongnized")
```

Comment:
As above, this function is way too long. The different mode clauses should be factored into their own functions.

Reply:
I have shortened this a bit (though maybe not enough). I'm a little hesitant to pull all the mode clauses into separate functions, but can do if it is felt that it will really help